### ISSUE TITLE
Unable to change new PNG picture on loop

### ISSUE THREAD

**Issue by Neothai** [kisvegabor/lvgl#6084]
### LVGL version

v9.1.0

### What happened?

Details of the hardware I use
MCU: ESP32-S3 Flash 16MB, PSRAM 8MB
Display: 3.5 inch TFT ILI9488 (16-bit parallel Communication type)

This project uses Arduino-IDE software developed.

Details of this project
Because my project needs to fetch a PNG image from the internet and display it on the display, I download the image and put it on LittleFS and then display the PNG image on the display using the PNG decoder within the library. Therefore, PNG images cannot be imported through LVGL's online image conversion tool.

I have to do this process every time. times when users request Therefore, it is necessary that I perform those processes on the loop of the program.

My main problem is I cannot change the new image even though the downloaded file data sent to LVGL is a new file data (with the same file name).
The nature of this problem is when I changed the image to a new one. But the same picture still appears on the display.

[ImageData.txt](https://github.com/lvgl/lvgl/files/15007045/ImageData.txt)

"data1" in the code is this image.
![landscape320240](https://github.com/lvgl/lvgl/assets/123863242/6bffdc7e-c0fc-42a0-af9b-3eb7d5b304af)

"data2" is this image.
![promptpay210x71](https://github.com/lvgl/lvgl/assets/123863242/86d12fdd-64a7-4c6e-9819-d4e0728ac9ae)

"data3" is this image.
![png_decoder_test](https://github.com/lvgl/lvgl/assets/123863242/df20dbb0-f27e-46ec-b082-bab808cbe5b3)

After powering on when executing this command (I'll call it "cmd1" for short.)
```cpp
if('a' == Serial.read()){
    if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old1------------------------");}
    imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new1");
    Serial.println("change img");
    base64_decodeFunc(data2, "/decoded_image.png"); //Convert images in Base64 format to PNG. Then store it in LittleFS with the file name "decoded_image.png"
    lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
    lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
  }
```
Result is
![IMG_20240417_142217_2](https://github.com/lvgl/lvgl/assets/123863242/6021e6cc-4bcb-4ae7-884e-8492e069c511)

and execute this command next(I'll call it "cmd3" for short.)
```cpp
if('c' == Serial.read()){
    if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old3------------------------");}
    imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new3");
    Serial.println("change img3");
    base64_decodeFunc(data3, "/decoded_image.png");
    lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
    lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
  }
```
Result is
![IMG_20240417_142217_2](https://github.com/lvgl/lvgl/assets/123863242/f0e73040-3c02-4df9-b1af-a5cb6c2cc45b)
The results remain the same as the first time.

But after powering on again and executing the "cmd3" command first, I get the following results.
![IMG_20240417_142231_2](https://github.com/lvgl/lvgl/assets/123863242/1811c3e5-1ec9-4389-9b83-c073f8d1ef5f)
and execute "cmd1" command next, Result is
![IMG_20240417_142231_3](https://github.com/lvgl/lvgl/assets/123863242/3840cbeb-5bc6-4baa-9db3-4f9e49bf39f7)

Is there any information you would like from me or what should I do next? Please tell me!😊

### How to reproduce?

```cpp
#include <lvgl.h>
#include "basicSystemSetting.h"
#include "arduino_base64.hpp"
#include "FS.h"
#include <LittleFS.h>
#include <ff.h>
#include <WiFi.h>

// The data for testing is as if this data were downloaded from the internet. But PNG is still in Base64 format.
// Located in the ImageData.txt file.

static void * lvglfsOpenFileHandler(lv_fs_drv_t * drv, const char * path, lv_fs_mode_t mode){
  LV_UNUSED(drv);
  const char *flags = "";

  if(mode == LV_FS_MODE_WR){flags = FILE_WRITE; Serial.println("lv_fs_wr");}
  else if(mode == LV_FS_MODE_RD){flags = FILE_READ; Serial.println("lv_fs_rd");}
  else if(mode == (LV_FS_MODE_WR | LV_FS_MODE_RD)){flags = FILE_WRITE;}

  Serial.println(path);
  File f = LittleFS.open(path, flags);
  if(!f){
    Serial.println("Failed to open file!");
    return NULL;
  }

  File *lf = new File{f};

  //make sure at the beginning
  //fp->seek(0);

  return (void *)lf;
}

static lv_fs_res_t lvglfsCloseFileHandler(lv_fs_drv_t *drv, void *file_p){
  LV_UNUSED(drv);

  Serial.println("lv_fs_closeF");

  File *fp = (File *)file_p;
  fp->close();

  delete (fp);  // when close
  return LV_FS_RES_OK;
}

static lv_fs_res_t lvglfsReadFileHandler(lv_fs_drv_t *drv, void *file_p, void *fileBuf, uint32_t btr, uint32_t *br){
  LV_UNUSED(drv);

  Serial.println("lv_fs_readData");

  File *fp = (File *)file_p;
  *br = fp->read((uint8_t *)fileBuf, btr);

  return (int32_t)(*br) < 0 ? LV_FS_RES_UNKNOWN : LV_FS_RES_OK;
}

static lv_fs_res_t lvglfsWriteFileHandler(lv_fs_drv_t *drv, void *file_p, const void *buf, uint32_t btw, uint32_t *bw){
  LV_UNUSED(drv);

  Serial.println("lv_fs_writeData");

  File *fp = (File *)file_p;
  *bw = fp->write((const uint8_t *)buf, btw);

  return (int32_t)(*bw) < 0 ? LV_FS_RES_UNKNOWN : LV_FS_RES_OK;
}

static lv_fs_res_t lvglfsSeekFileHandler(lv_fs_drv_t *drv, void *file_p, uint32_t pos, lv_fs_whence_t whence){
  LV_UNUSED(drv);

  File *fp = (File *)file_p;

  SeekMode mode;
  if (whence == LV_FS_SEEK_SET){mode = SeekSet;}
  else if (whence == LV_FS_SEEK_CUR){mode = SeekCur;}
  else if (whence == LV_FS_SEEK_END){mode = SeekEnd;}

  fp->seek(pos, mode);

  return LV_FS_RES_OK;
}

static lv_fs_res_t lvglfsTellFileHandler(lv_fs_drv_t *drv, void *file_p, uint32_t *pos_p){
  LV_UNUSED(drv);

  File *fp = (File *)file_p;
  *pos_p = fp->position();

  return LV_FS_RES_OK;
}

void base64_decodeFunc(String in, const char* fn){
  char *input = (char *)ps_malloc(in.length() + 1);    
  strcpy(input, in.c_str());
    
  // Decode Base64
  uint8_t *output = (uint8_t *)ps_malloc(base64::decodeLength(input));
  base64::decode(input, output);
  delay(100); Serial.println("decode base64 OK!");

  File file;
  if(LittleFS.exists(fn)) {
    Serial.println("exists, remove");
    LittleFS.remove(fn);    
  }
  file = LittleFS.open(fn, FILE_WRITE);
  if(file){
     Serial.println("Writing...");
     file.write(output, base64::decodeLength(input));
     Serial.println("OK!");
  }
  file.close();
  Serial.println("free mem");

  free(input);
  free(output); 
}

lv_obj_t * imgobj_qr1;

void setup(){
  Serial.begin(115200);   
  
  lvglDisplayConfig();

  if(!LittleFS.begin(0)){
    Serial.println("LittleFS Mount Failed");
    while(!LittleFS.begin(0));
  }

  Serial.println("LittleFS OK");

  delay(500);

  lv_fs_stdio_init();
  
  static lv_fs_drv_t drv;                           /*Needs to be static or global*/
  lv_fs_drv_init(&drv);                             /*Basic initialization*/

  drv.letter =        'C';                          /*An uppercase letter to identify the drive */
  drv.cache_size =    (128*1024);                        /*Cache size for reading in bytes. 0 to not cache.*/

  drv.ready_cb =      NULL;                         /*Callback to tell if the drive is ready to use */
  drv.open_cb =       lvglfsOpenFileHandler;        /*Callback to open a file */
  drv.close_cb =      lvglfsCloseFileHandler;       /*Callback to close a file */
  drv.read_cb =       lvglfsReadFileHandler;        /*Callback to read a file */
  drv.write_cb =      lvglfsWriteFileHandler;       /*Callback to write a file */
  drv.seek_cb =       lvglfsSeekFileHandler;        /*Callback to seek in a file (Move cursor) */
  drv.tell_cb =       lvglfsTellFileHandler;        /*Callback to tell the cursor position  */

  drv.dir_open_cb =   NULL;                         /*Callback to open directory to read its content */
  drv.dir_read_cb =   NULL;                         /*Callback to read a directory's content */
  drv.dir_close_cb =  NULL;                         /*Callback to close a directory */

  //drv.user_data = my_user_data;                   /*Any custom data if required*/
  lv_fs_drv_register(&drv);                         /*Finally register the drive*/
  Serial.println("drv_register OK!");

  imgobj_qr1 = lv_image_create(lv_screen_active());    
}
unsigned long lvgl_tick_millis;

void loop(){
  lv_timer_handler(); /* let the UI do its work */

  if('a' == Serial.read()){
    if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old1------------------------");}
    imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new1");
    Serial.println("change img");
    base64_decodeFunc(data2, "/decoded_image.png"); //Convert images in Base64 format to PNG. Then store it in LittleFS with the file name "decoded_image.png"
    lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
    lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
  }
  
  if('b' == Serial.read()){
    if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old2------------------------");}
    imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new2");
    Serial.println("change img2");
    base64_decodeFunc(data1, "/decoded_image2.png");
    lv_image_set_src(imgobj_qr1, "C:/decoded_image2.png");
    lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
  }
  
  if('c' == Serial.read()){
    if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old3------------------------");}
    imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new3");
    Serial.println("change img3");
    base64_decodeFunc(data3, "/decoded_image.png");
    lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
    lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
  }

  unsigned long tick_millis = millis() - lvgl_tick_millis;
  lvgl_tick_millis = millis();
  lv_tick_inc(tick_millis);
  delay(2);
}
```

**Comment by kisvegabor** [comment]
Does the UI still work after changing the image?

Please try these:
- call `lv_obj_invalidate(imgobj_qr1)`
- call `lv_image_cache_drop("C:/decoded_image.png");`
- set `LV_CACHE_DEF_SIZE 0` in `lv_conf.h`
- try saving the image with a different name

**Comment by Neothai** [comment]
Thank you for the good advice. I did the test as you suggested and got results like this.

> Does the UI still work after changing the image

Yes, the UI and other objects still work normally.
___
> call `lv_obj_invalidate(imgobj_qr1)`

I have tried doing this
``` 
...
lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);

lv_obj_invalidate(imgobj_qr1); // <---
```
But still nothing has changed.
____

> call `lv_image_cache_drop("C:/decoded_image.png");`

**It works**, but needs to call `lv_image_cache_drop("C:/decoded_image.png");` before creating an image like this.
```
lv_image_cache_drop("C:/decoded_image.png"); //<--
if(lv_obj_is_valid(imgobj_qr1)){lv_obj_delete_async(imgobj_qr1); Serial.println("del old1------------------------");}
imgobj_qr1 = lv_image_create(lv_screen_active()); Serial.println("create new1");
...
lv_image_set_src(imgobj_qr1, "C:/decoded_image.png");
lv_obj_align(imgobj_qr1, LV_ALIGN_CENTER, 0, 0);
```
____
> set `LV_CACHE_DEF_SIZE` 0 in lv_conf.h`

I originally set this to
```
#define LV_MEM_SIZE                   (3072U * 1024U) // Allocate on PSRAM

#define LV_CACHE_DEF_SIZE             (1500U * 1024U)
#define LV_IMAGE_HEADER_CACHE_DEF_CNT (512U * 1024U)
```
When changed to
```
#define LV_MEM_SIZE                   (3072U * 1024U) // Allocate on PSRAM

#define LV_CACHE_DEF_SIZE             0
#define LV_IMAGE_HEADER_CACHE_DEF_CNT 0
```
It works too. But it takes a very long time to draw and the file is constantly being read (for large images).
____

> try saving the image with a different name

It works too (when `LV_CACHE_DEF_SIZE` and `LV_IMAGE_HEADER_CACHE_DEF_CNT` are not 0), but images in the old cache are not cleared when changing to new images. Then the cache will be full when changing the image to a new one within 3-4 times.
____
In conclusion, I think your suggestion **call `lv_image_cache_drop("C:/decoded_image.png");`** is the best solution.

And this would be great if in the documentation https://docs.lvgl.io/master/libs/lodepng.html it was suggested to add `lv_image_cache_drop(imageSrc);` before loading new images.

Thanks again for the great solution to this problem. 😊

**Comment by kisvegabor** [comment]
Awesome! :tada: 

Actually it's related to images in general, because LVGL caches images by file name (or pointer in case of variables). So if you change the content of the image you need to tell LVGL to re-cache it.

@W-Mai I just saw that the docs of the image caching is outdated [here](https://docs.lvgl.io/overview/image.html#image-caching). Do you have time to update it?
